testing/proxytest/https.go

```diff
@@ -0,0 +1,220 @@
+	log.Debug("hijacked request")
+
+	// ==================== CONNECT accepted, let the client know
+	_, err = clientCon.Write([]byte("HTTP/1.1 200 Connection established\r\n\r\n"))
+	if err != nil {
+		p.http500Error(clientCon, "failed to send 200-OK after CONNECT", err, log)
+		return
+	}
+
+	// ==================== TLS handshake
+	// client will proceed to perform the TLS handshake with the "target",
+	// which we're impersonating.
+
+	// generate a TLS certificate matching the target's host
+	cert, err := p.newTLSCert(r.URL)
+	if err != nil {
+		p.http500Error(clientCon, "failed generating certificate", err, log)
+		return
+	}
+
+	tlscfg := p.TLS.Clone()
+	tlscfg.Certificates = []tls.Certificate{*cert}
+	clientTLSConn := tls.Server(clientCon, tlscfg)
+	defer clientTLSConn.Close()
+	err = clientTLSConn.Handshake()
+	if err != nil {
+		p.http500Error(clientCon, "failed TLS handshake with client", err, log)
+		return
+	}
+
+	clientTLSReader := bufio.NewReader(clientTLSConn)
+
+	notEOF := func(r *bufio.Reader) bool {
+		_, err = r.Peek(1)
+		return !errors.Is(err, io.EOF)
+	}
+	// ==================== Handle the actual request
+	for notEOF(clientTLSReader) {
+		// read request from the client sent after the 1s CONNECT request
+		req, err := http.ReadRequest(clientTLSReader)
+		if err != nil {
+			p.http500Error(clientTLSConn, "failed reading client request", err, log)
+			return
+		}
+
+		// carry over the original remote addr
+		req.RemoteAddr = r.RemoteAddr
+
+		// the read request is relative to the host from the original CONNECT
+		// request and without scheme. Therefore, set them in the new request.
+		req.URL, err = url.Parse("https://" + r.Host + req.URL.String())
+		if err != nil {
+			p.http500Error(clientTLSConn, "failed reading request URL from client", err, log)
+			return
+		}
+		cleanUpHeaders(req.Header)
+
+		// now the request is ready, it can be altered just as an HTTP request
+		// can.
+		resp, err := p.processRequest(req)
+		if err != nil {
+			p.http500Error(clientTLSConn, "failed performing request to target", err, log)
+			return
+		}
+
+		// Send response from target to client
+		// 1st - the status code
+		_, err = clientTLSConn.Write([]byte("HTTP/1.1 " + resp.Status + "\r\n"))
+		if err != nil {
+			p.http500Error(clientTLSConn, "failed writing response status line", err, log)
+			return
+		}
+
+		// 2nd - the headers
+		if err = resp.Header.Write(clientTLSConn); err != nil {
+			p.http500Error(clientTLSConn, "failed writing TLS response header", err, log)
+			return
+		}
+
+		// 3rd - indicates the headers are done and the body will follow
+		if _, err = clientTLSConn.Write([]byte("\r\n")); err != nil {
+			p.http500Error(clientTLSConn, "failed writing TLS header/body separator", err, log)
+			return
+		}
+
+		// copy the body else
+		_, err = io.CopyBuffer(clientTLSConn, resp.Body, make([]byte, 4096))
+		if err != nil {
+			p.http500Error(clientTLSConn, "failed writing response body", err, log)
+			return
+		}
```

Comment:
Can't we just do `resp.Write(clientTLSConn)`? Doesn't look like we're customizing the output in any way.